aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/MutationQueue.java

```diff
@@ -0,0 +1,317 @@
+package com.amplifyframework.datastore.syncengine;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.amplifyframework.core.model.Model;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Queue;
+import io.reactivex.Observable;
+import io.reactivex.functions.Function;
+
+/*
+ * The {@link MutationQueue} is a LinkedHashMap like container , the goal of using this container is to
+ * achieve O(1) time complexity for both getting a {@link PendingMutation} and update an existing mutation with
+ * valid id.
+ *
+ * MutationQueue is implementing the Queue interface and provide most of the queue operations,
+ */
+public class MutationQueue implements Queue<PendingMutation<? extends Model>> {
```

Comment:
Can be a `final` class, most likely.

The `Queue<...>` ends up bringing in a lot of methods, actually. Now that you've investigate that route, do you think they add value? Or not? It was an interesting idea, to implement the `Queue` interface, but perhaps it is overkill. Particularly if many will just throw an exception.

Reply:
@jamesonwilliams  will revise according to your comments, you are actually right that using a Queue interface is a overkill, especially we cant even reference this object in MutationOutBox as a Queue because we need the findMutationById() method to be exposed which is not a standard Queue operation,

Reply:
@jamesonwilliams made needed change in the latest commit :)